File: InterviewBit/linkedlists/rotatelist.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode* Solution::rotateRight(ListNode* head, int k){

      if(head == nullptr || head->next == nullptr) return head;

      int stop,length = 0;
      ListNode *temp,*newHead,*cur = head;
      while(cur){ length++; cur = cur->next; }

      k = k%length;
      if(!k) return head;

      stop = length-k-1;
      cur = head;
      while(stop){
          cur = cur->next;
          stop--;
      }

      newHead =  cur->next;
      cur->next = nullptr;
      cur = newHead;
      while(cur->next) cur = cur->next;
      cur->next = head;
      return newHead;

}

File: LeetCode/Java/src/main/java/RotateList.java
import common.ListNode;

import java.util.List;

public class RotateList {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null) return null;

        ListNode current = head;
        int length = 0;
        while(current != null) {
            length += 1;
            current = current.next;
        }

        k %= length;
        ListNode forward = head;
        current = head;
        while(k >0) {
            forward = forward.next;
            k--;
        }
        while(forward.next != null) {
            current = current.next;
            forward = forward.next;
        }

        forward.next = head;
        ListNode newHead = current.next;
        current.next = null;
        return newHead;
    }
}
